Data/HashSet.hs

```diff
@@ -4,25 +4,95 @@
 #endif
 
 ------------------------------------------------------------------------
--- |
--- Module      :  Data.HashSet
--- Copyright   :  2011 Bryan O'Sullivan
--- License     :  BSD-style
--- Maintainer  :  johan.tibell@gmail.com
--- Stability   :  provisional
--- Portability :  portable
---
--- A set of /hashable/ values.  A set cannot contain duplicate items.
--- A 'HashSet' makes no guarantees as to the order of its elements.
---
--- The implementation is based on /hash array mapped trie/.  A
--- 'HashSet' is often faster than other tree-based set types,
--- especially when value comparison is expensive, as in the case of
--- strings.
---
--- Many operations have a average-case complexity of /O(log n)/.  The
--- implementation uses a large base (i.e. 16) so in practice these
--- operations are constant time.
+{-|
+Module      :  Data.HashSet
+Copyright   :  2011 Bryan O'Sullivan
+License     :  BSD-style
+Maintainer  :  johan.tibell@gmail.com
+Stability   :  provisional
+Portability :  portable
+
+= Introduction
+
+'HashSet' allows you to store /unique/ elements, providing efficient
+insertion, lookups, and deletion. If you are storing sets of @Int@s consider
+using the @Data.IntSet@ from the
+<https://hackage.haskell.org/packages/containers containers> package.
+A 'HashSet' makes no guarantees as to the order of its elements.
+
+== Examples
+
+All the examples below assume @HashSet@ is imported qualified, and uses the following @dataStructures@ set.
+
+>>> import qualified Data.HashSet as HashSet
+>>> let dataStructures = HashSet.fromList ["Set", "Map", "Graph", "Sequence"]
+
+=== Basic Operations
+
+Check membership in a set:
+
+>>> -- Check if "Map" and "Trie" are in the set of data structures.
+>>> HashSet.member "Map" dataStructures
+True
+>>> HashSet.member "Trie" dataStructures
+False
+
+Add a new entry to the set:
+
+>>> let moreDataStructures = HashSet.insert "Trie" dataStructures
+>>> HashSet.member \"Trie\" moreDataStructures
+> True
+
+Remove the @\"Graph\"@ entry from the set of data structures.
+
+>>> let fewerDataStructures = HashSet.delete "Graph" dataStructures
+>>> HashSet.toList fewerDataStructures
+["Map","Set","Sequence"]
+
+
+Create a new set and combine it with our original set.
+
+>>> let unorderedDataStructures = HashSet.fromList ["HashSet", "HashMap"]
+>>> HashSet.union dataStructures unorderedDataStructures
+fromList ["Map","HashSet","Graph","HashMap","Set","Sequence"]
+
+=== Using custom data with HashSet
+
+To create a @HashSet@ of your custom type, the type must have instances for 'Eq' and
+'Hashable'. The 'Hashable' typeclass is defined in the
+<https://hackage.haskell.org/packages/hashable hashable> package and the
+recommended way to define the instance is using generics, for which you'll need
+the @DeriveGeneric@ GHC extension.
+
+Note: You'll need to enable the @DeviceGeneric@ language extension, either via
+@:set -XDeriveGeneric@ in GHCi or @\{\-\# DeriveGeneric \#\-\}@ in your Haskell source
+code.
```

Comment:
I think it's slightly unfortunate to go into so much detail here when the canonical documentation on this is in `hashable`: http://hackage.haskell.org/package/hashable-1.3.0.0/docs/Data-Hashable.html#g:3

Maybe just add a link?!

Reply:
No need to add anything. If @m-renaud wants that in the ReadTheDocs stuff, that's his prerogative, but it doesn't belong in the Haddocks.

Reply:
@treeowl I think the current idea is not to add a ReadTheDocs page, but instead to include some examples and a bit of introductory documentation in the main haddocks.

I think there's some value in including an example of using u-c with custom datatypes, but I mostly feel that it shouldn't duplicate documentation that is or should be in hashable instead…

Reply:
Yeah, my rationale for including a bit more than a link out was to prevent documentation hops, especially given that u-c is likely one of the first packages that someone new to the language may encounter. It does add a _bit_ more content to the docs but having a full working example in one place without having to piece things together can be very useful.

Additionally, the canonical documentation doesn't mention what to do if you're in ghci, so I thought including it here would be helpful as well for those not familiar with extensions. I don't think we should assume that anyone reading these docs knows that `:set -XExtensionName` is how to enable extensions in ghci; yes they're bound to figure it out eventually the more time you spend in the Haskell ecosystem, but one extra sentence to avoid a stackoverflow or reddit post seems like a good trade-off to me.

@sjakobi @treeowl please either :+1:  or :-1: this comment:

- :+1: - rationale sounds good, lets leave this
- :-1: - still don't agree, please remove details

Reply:
I'm very happy with the section in its current state.

> u-c is likely one of the first packages that someone new to the language may encounter

This is surprising to me! Why do you think new users will encounter it so soon? u-c always seemed fairly niche compared to `containers` to me.